src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -594,19 +594,33 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
             static async Task<T> CompleteAsync(Task task, string hostName, long startingTimestamp)
             {
                 _ = NameResolutionTelemetry.Log.BeforeResolution(hostName);
-                T? result = null;
+                string? errorType = null;
                 try
                 {
-                    result = await ((Task<T>)task).ConfigureAwait(false);
-                    return result;
+                    return await ((Task<T>)task).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    errorType = GetErrorType(ex);
+                    throw;
                 }
                 finally
                 {
-                    NameResolutionTelemetry.Log.AfterResolution(hostName, startingTimestamp, successful: result is not null);
+                    NameResolutionTelemetry.Log.AfterResolution(hostName, startingTimestamp, errorType);
                 }
             }
         }
 
+        private static string GetErrorType(Exception exception) => (exception as SocketException)?.SocketErrorCode switch
+        {
+            SocketError.HostNotFound => "host_not_found",
+            SocketError.TryAgain => "try_again",
+            SocketError.AddressFamilyNotSupported => "address_family_not_supported",
+            SocketError.NoRecovery => "no_recovery",
+
+            _ => exception.GetType().Name
+        };
```

Comment:
I picked the codes we can realistically get from the PAL by looking at our [mapping in the Unix PAL](https://github.com/dotnet/runtime/blob/94f212275b2f51ca67025d677d7d5c5bc75f670f/src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Unix.cs#L23-L46), and [`getaddrinfo` return values on Windows](https://learn.microsoft.com/en-us/windows/win32/api/ws2tcpip/nf-ws2tcpip-getaddrinfo#return-value).
There are some other codes that may theoretically occur in practice. Ie. seeing `InvalidArgument` , `TypeNotFound` or `SocketNotSupported` means a programming error in our PAL if I'm not mistaken.